File: test/permissions/test_permissions.f90
program main

use, intrinsic :: iso_fortran_env, only : stderr=>error_unit
use filesystem

implicit none

character(9) :: p


character(*), parameter :: reada="readable.txt", noread="not-readable.txt", nowrite="not-writable.txt"

p = get_permissions("")
if(len_trim(p) /= 0) then
    write(stderr, '(a)') "get_permissions('') should be empty: " // p
    error stop
endif

!> readable
call touch(reada)
call set_permissions(reada, readable=.true.)

p = get_permissions(reada)
print '(a)', "Permissions for " // trim(reada)// ": "// p

if (len_trim(p) == 0) error stop "get_permissions('"//trim(reada)//"') should not be empty"

if (p(1:1) /= "r") then
    write(stderr, '(a)') "ERROR: test_exe: "//trim(reada)//" should be readable"
    error stop
endif
if(.not. is_readable(reada)) error stop "test_exe: "//trim(reada)//" should be readable"

if(.not. exists(reada)) error stop "test_exe: "//trim(reada)//" should exist"

if(.not. is_file(reada)) error stop "test_exe: "//trim(reada)//" should be a file"

!! for Ffilesystem, even non-readable files "exist" and are "is_file"

call touch(noread)
call set_permissions(noread, readable=.false.)

p = get_permissions(noread);
print '(a)', "Permissions for " // trim(noread)// ": "// p

if (len_trim(p) == 0) error stop "get_permissions('"//trim(noread)//"') should not be empty"

if (index(p, "r") /= 0) then
    write(stderr, '(a)') "XFAIL:test_exe: "//trim(noread)//" should not be readable"
else
if (is_readable(noread)) error stop "test_exe: "//trim(noread)//" should not be readable"

if (.not. exists(noread)) error stop "test_exe: "//trim(noread)//" should exist"

if (.not. is_file(noread)) error stop "test_exe: "//trim(noread)//" should be a file"
endif

!> writable
if(.not. is_file(nowrite)) call touch(nowrite)
call set_permissions(nowrite, writable=.false.)

if (.not. is_writable(".")) error stop "test_exe: . should be writable"

p = get_permissions(nowrite);
print '(a)', "Permissions for " // trim(nowrite)// ": "// p

if (len_trim(p) == 0) error stop "get_permissions('"//trim(nowrite)//"') should not be empty"

if (index(p, "w") /= 0) then
    write(stderr, '(a)') "test_exe: "//trim(nowrite)//" should not be writable"
    error stop 77
endif
if(is_writable(nowrite)) error stop "test_exe: "//trim(nowrite)//" should not be writable"

if (.not. exists(nowrite)) error stop "test_exe: "//trim(nowrite)//" should exist"

if (.not. is_file(nowrite)) error stop "test_exe: "//trim(nowrite)//" should be a file"

call remove(reada)
call remove(noread)
call remove(nowrite)

end program

File: test/permissions/test_permissions.cpp
#include <iostream>
#include <cstdlib>
#include <string>

#include "ffilesystem.h"

#ifdef _MSC_VER
#include <crtdbg.h>
#endif


[[noreturn]] void err(std::string_view m){
    std::cerr << "ERROR: " << m << "\n";
    std::exit(EXIT_FAILURE);
}


int main()
{

#ifdef _MSC_VER
  _CrtSetReportMode(_CRT_ASSERT, _CRTDBG_MODE_FILE);
  _CrtSetReportFile(_CRT_ASSERT, _CRTDBG_FILE_STDERR);
  _CrtSetReportMode(_CRT_WARN, _CRTDBG_MODE_FILE);
  _CrtSetReportFile(_CRT_WARN, _CRTDBG_FILE_STDERR);
  _CrtSetReportMode(_CRT_ERROR, _CRTDBG_MODE_FILE);
  _CrtSetReportFile(_CRT_ERROR, _CRTDBG_FILE_STDERR);
#endif

std::string p = Ffs::get_permissions("");
if(p.length() != 0)
    err("get_permissions('') should be empty, got: " + p);

std::string read = "readable.txt";
std::string noread = "nonreadable.txt";
std::string nowrite = "nonwritable.txt";

Ffs::touch(read);
Ffs::set_permissions(read, 1, 0, 0);

p = Ffs::get_permissions(read);
std::cout << "Permissions for " << read << ": " << p << "\n";

if(p.length() == 0)
    err("get_permissions('" + read + "') should not be empty");

if(!Ffs::is_readable(read))
    err(read + " should be readable");

if(!Ffs::exists(read))
    err(read + " should exist");

if(!Ffs::is_file(read))
    err(read + " should be a file");

// for Ffilesystem, even non-readable files "exist" and are "is_file"
Ffs::touch(noread);
Ffs::set_permissions(noread, -1, 0, 0);

p = Ffs::get_permissions(noread);
std::cout << "Permissions for " << noread << ": " << p << "\n";

if(p.length() == 0)
    err(noread + "') should not be empty");

if(Ffs::is_readable(noread)){
    std::cerr << "XFAIL: test_exe: " << noread << " should not be readable\n";
} else {

if(!Ffs::exists(noread))
    err(noread + " should exist");

if(!Ffs::is_file(noread))
    err(noread + " should be a file");
}
// writable
if(!Ffs::is_file(nowrite))
  Ffs::touch(nowrite);
Ffs::set_permissions(nowrite, 0, -1, 0);

std::cout << "Permissions for " << nowrite << ": " << Ffs::get_permissions(nowrite) << "\n";

if(Ffs::is_writable(nowrite)){
    std::cerr << "ERROR: test_exe: " << nowrite << " should not be writable\n";
    return 77;
}

if(!Ffs::exists(nowrite))
    err(nowrite + " should exist");

if(!Ffs::is_file(nowrite))
    err(nowrite + " should be a file");

return EXIT_SUCCESS;
}
